pkg/checks/traceroute/check.go

```diff
@@ -0,0 +1,195 @@
+	Port uint16 `json:"port" yaml:"port" mapstructure:"port"`
+}
+
+func NewCheck() checks.Check {
+	return &Traceroute{
+		CheckBase: checks.CheckBase{
+			Mu:       sync.Mutex{},
+			DoneChan: make(chan struct{}, 1),
+		},
+		config:     Config{},
+		traceroute: TraceRoute,
+		metrics:    newMetrics(),
+	}
+}
+
+type Traceroute struct {
+	checks.CheckBase
+	config     Config
+	traceroute tracerouteFactory
+	metrics    metrics
+}
+
+type tracerouteConfig struct {
+	Dest    string
+	Port    int
+	Timeout int
+	MaxHops int
+	Rc      helper.RetryConfig
+}
+type tracerouteFactory func(ctx context.Context, cfg tracerouteConfig) (map[int][]Hop, error)
+
+type result struct {
+	// The minimum number of hops required to reach the target
+	MinHops int
+	// The path taken to the destination
+	Hops map[int][]Hop
+}
+
+// Run runs the check in a loop sending results to the provided channel
+func (tr *Traceroute) Run(ctx context.Context, cResult chan checks.ResultDTO) error {
+	ctx, cancel := logger.NewContextWithLogger(ctx)
+	defer cancel()
+	log := logger.FromContext(ctx)
+
+	log.Info("Starting traceroute check", "interval", tr.config.Interval.String())
+	for {
+		select {
+		case <-ctx.Done():
+			log.Error("Context canceled", "error", ctx.Err())
+			return ctx.Err()
+		case <-tr.DoneChan:
+			return nil
+		case <-time.After(tr.config.Interval):
+			res := tr.check(ctx)
+			tr.metrics.MinHops(res)
+			cResult <- checks.ResultDTO{
+				Name: tr.Name(),
+				Result: &checks.Result{
+					Data:      res,
+					Timestamp: time.Now(),
+				},
+			}
+			log.Debug("Successfully finished traceroute check run")
+		}
+	}
+}
+
+// GetConfig returns the current configuration of the check
+func (tr *Traceroute) GetConfig() checks.Runtime {
+	tr.Mu.Lock()
+	defer tr.Mu.Unlock()
+	return &tr.config
+}
+
+func (tr *Traceroute) check(ctx context.Context) map[string]result {
+	res := make(map[string]result)
+	log := logger.FromContext(ctx)
+
+	type internalResult struct {
+		addr string
+		res  result
+	}
+
+	cResult := make(chan internalResult, len(tr.config.Targets))
+	var wg sync.WaitGroup
+
+	wg.Add(len(tr.config.Targets))
+	for _, t := range tr.config.Targets {
+		go func(t Target) {
+			defer wg.Done()
+			l := log.With("target", t.Addr)
+			l.Debug("Running traceroute")
+
+			start := time.Now()
+			trace, err := tr.traceroute(ctx, tracerouteConfig{
+				Dest:    t.Addr,
+				Port:    int(t.Port),
+				Timeout: int(tr.config.Timeout / time.Millisecond),
+				MaxHops: tr.config.MaxHops,
+				Rc:      tr.config.Retry,
+			})
+			elapsed := time.Since(start)
+			if err != nil {
+				l.Error("Error running traceroute", "error", err)
+			}
+
+			tr.metrics.CheckDuration(t.Addr, elapsed)
+
+			l.Debug("Ran traceroute", "result", trace, "duration", elapsed)
+			r := result{
+				Hops:    trace,
+				MinHops: tr.config.MaxHops,
+			}
+
+			for i, hops := range trace {
+				for _, hop := range hops {
+					if hop.Reached && hop.Ttl < r.MinHops {
+						r.MinHops = i
```

Comment:
~~r.MinHps should be i-1~~
 
edit: ok since ttl is obviously starting at 1 ✅

Reply:
~~Is there a possibility to have multiple hops per hop since hops is a slice as well?~~

edit: sure! retries are listed in on hop.

maybe the wording here is not correct. would it be better to use: `hop` instead of `hops` and `retry` instead of `hop`